File: ga/glm4_result_2024_6/result/1968D_glm/27_1968D.c
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut lines = input.lines();
    
    let t: usize = lines.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let mut values = lines.next().unwrap().split_whitespace().map(|s| s.parse::<usize>().unwrap());
        let n = values.next().unwrap();
        let k = values.next().unwrap();
        let a = values.next().unwrap();
        let b = values.next().unwrap();
        
        let mut arr = [0; 200001];
        let mut aa = [0; 200001];
        for i in 1..=n {
            arr[i] = lines.next().unwrap().parse().unwrap();
        }
        for i in 1..=n {
            aa[i] = lines.next().unwrap().parse().unwrap();
        }

        let mut p1 = 0;
        let mut p2 = 0;
        let mut a = a;
        let mut b = b;
        for _ in 0..k {
            p1 = p1.max(1 * aa[a] + p1);
            p1 += aa[a];
            a = arr[a];
            
            p2 = p2.max(1 * aa[b] + p2);
            p2 += aa[b];
            b = arr[b];
        }
        
        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/gpt3_5_result_2024_6/result/1968D_cpp_gpt/27_1968D.rs
use std::io::{self, BufRead};

fn calculate(p: &Vec<usize>, a: &Vec<i64>, mut pos: usize, mut k: i64) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    
    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + k * a[pos]);
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    
    mx
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines().map(Result::unwrap);
    
    let t: usize = lines.next().unwrap().parse().unwrap();
    
    for _ in 0..t {
        let line: Vec<usize> = lines.next().unwrap().split_whitespace()
                                    .map(|x| x.parse().unwrap()).collect();
        let n = line[0];
        let k = line[1] as i64;
        let ps1 = line[2];
        let ps2 = line[3];
        
        let p: Vec<usize> = lines.next().unwrap().split_whitespace()
                                .map(|x| x.parse().unwrap() - 1).collect();
        let a: Vec<i64> = lines.next().unwrap().split_whitespace()
                                .map(|x| x.parse().unwrap()).collect();
        
        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);
        
        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}